File: packages/node/src/retry.ts
/**
 * Retry decisions for the VendorVal API.
 *
 * Retryable:
 *   - network failures (no response received)
 *   - HTTP 408 Request Timeout
 *   - HTTP 429 Too Many Requests (honors `retry-after` / `x-ratelimit-reset`)
 *   - HTTP 5xx
 */

export interface RetryDecision {
  retry: boolean;
  /** Sleep before retrying, in ms. */
  delayMs: number;
}

const BASE_DELAY_MS = 500;
const MAX_DELAY_MS = 30_000;

export function shouldRetryStatus(status: number): boolean {
  if (status === 408 || status === 429) return true;
  return status >= 500 && status <= 599;
}

export function computeBackoffMs(attempt: number, jitter = Math.random()): number {
  // Exponential: 0.5s, 1s, 2s, 4s, ... capped.
  const exp = Math.min(MAX_DELAY_MS, BASE_DELAY_MS * 2 ** attempt);
  // Full jitter in [exp/2, exp).
  return Math.floor(exp / 2 + jitter * (exp / 2));
}

export function decideRetryFromHeaders(
  attempt: number,
  status: number,
  headers: Headers,
): RetryDecision {
  if (!shouldRetryStatus(status)) {
    return { retry: false, delayMs: 0 };
  }

  if (status === 429) {
    const retryAfter = headerSeconds(headers.get("retry-after"));
    if (retryAfter !== null) {
      return { retry: true, delayMs: Math.min(MAX_DELAY_MS, retryAfter * 1000) };
    }
    const reset = headers.get("x-ratelimit-reset");
    if (reset) {
      // The header may be either an HTTP-date string (Date.parse-compatible)
      // or a bare unix-epoch-seconds integer. Date.parse returns NaN on the
      // latter, so detect numeric input first.
      const epochMs = /^\d+$/.test(reset.trim())
        ? Number.parseInt(reset, 10) * 1000
        : Date.parse(reset);
      if (Number.isFinite(epochMs)) {
        return {
          retry: true,
          delayMs: Math.min(MAX_DELAY_MS, Math.max(0, epochMs - Date.now())),
        };
      }
    }
  }

  return { retry: true, delayMs: computeBackoffMs(attempt) };
}

function headerSeconds(raw: string | null): number | null {
  if (!raw) return null;
  const n = Number.parseInt(raw, 10);
  if (Number.isFinite(n)) return n;
  const epoch = Date.parse(raw);
  if (Number.isFinite(epoch)) return Math.max(0, Math.ceil((epoch - Date.now()) / 1000));
  return null;
}

File: packages/python/src/vendorval_sdk/_retry.py
"""Retry policy."""

from __future__ import annotations

import random
from datetime import UTC

import httpx

from ._errors import parse_retry_after

_BASE_DELAY = 0.5
_MAX_DELAY = 30.0


def should_retry_status(status: int) -> bool:
    if status in (408, 429):
        return True
    return 500 <= status <= 599


def compute_backoff(attempt: int) -> float:
    exp = min(_MAX_DELAY, _BASE_DELAY * (2 ** attempt))
    # Full jitter in [exp/2, exp).
    return float(exp / 2 + random.random() * (exp / 2))


def decide_retry(
    *,
    attempt: int,
    status: int,
    headers: httpx.Headers,
) -> float | None:
    """Return delay in seconds if we should retry, else None."""
    if not should_retry_status(status):
        return None
    if status == 429:
        retry_after = parse_retry_after(headers.get("retry-after"))
        if retry_after is not None:
            return min(_MAX_DELAY, retry_after)
        reset = headers.get("x-ratelimit-reset")
        if reset:
            try:
                from datetime import datetime

                ts = datetime.fromisoformat(reset.replace("Z", "+00:00"))
                delta = (ts - datetime.now(UTC)).total_seconds()
                return max(0.0, min(_MAX_DELAY, delta))
            except ValueError:
                pass
    return compute_backoff(attempt)
